src/lib.rs

```diff
@@ -1021,59 +1023,95 @@ fn get_docker_for_tcp(tcp_host_str: String) -> Docker {
             tcp_host_str
         };
 
-        Docker {
+        Ok(Docker {
             transport: Transport::EncryptedTcp {
                 client: Client::builder()
                     .build(HttpsConnector::with_connector(http, connector).unwrap()),
                 host: tcp_host_str,
             },
-        }
+        })
     } else {
-        Docker {
+        Ok(Docker {
             transport: Transport::Tcp {
                 client: Client::builder().build(http),
                 host: tcp_host_str,
             },
-        }
+        })
     }
 }
 
 #[cfg(not(feature = "tls"))]
-fn get_docker_for_tcp(tcp_host_str: String) -> Docker {
+fn get_docker_for_tcp<S>(tcp_host_str: S) -> Result<Docker>
+where
+    S: Into<String>,
+{
     let http = get_http_connector();
-    Docker {
+    Ok(Docker {
         transport: Transport::Tcp {
             client: Client::builder().build(http),
             host: tcp_host_str,
         },
+    })
+}
+
+impl Default for Docker {
+    /// Returns a Docker instance connected to a unix socket at [DEFAULT_SOCKET_PATH](DEFAULT_SOCKET_PATH)
+    fn default() -> Self {
+        Docker::unix(DEFAULT_SOCKET_PATH)
     }
 }
 
 // https://docs.docker.com/reference/api/docker_remote_api_v1.17/
 impl Docker {
+    /// Constructs a new Docker instance based on uri.
+    ///
+    /// uri can be either a unix or tcp socket, for example `unix:///run/docker.sock` or for tcp
+    /// `tcp://127.0.0.1:80`.
+    ///
+    /// Valid schemes are: `tcp`, `http`, `https`, `unix`
+    pub fn new<S>(uri: S) -> Result<Docker>
+    where
+        S: Into<String>,
+    {
+        let uri_str = uri.into();
+
+        #[allow(clippy::single_match)] // this match is much nicer than two separate if lets
+        match uri_str.strip_prefix("unix://") {
+            #[cfg(feature = "unix-socket")]
+            Some(path) => return Ok(Docker::unix(path)),
+            #[cfg(not(feature = "unix-socket"))]
+            Some(_) => return Err(Error::UnsupportedScheme { scheme: "unix" }),
+            _ => {}
+        }
+
+        let uri = uri_str.parse::<Uri>()?;
+
+        if let Some(scheme) = uri.scheme() {
+            match scheme.as_str() {
+                "tcp" | "http" | "https" => Docker::tcp(uri_str),
+                s => Err(Error::UnsupportedScheme { scheme: s.into() }),
+            }
+        } else {
+            Err(Error::UnsupportedScheme { scheme: "".into() })
+        }
+    }
```

Comment:
Given you are a fan of `match` vs `if let`, what do you think of https://github.com/softprops/shiplift/commit/879a5895496edf1b61cdec22d4404e5c0109f14a?

Reply:
That looks awesome! Will definitely use it.